### ISSUE TITLE
cant Request in callbacks v2 and reply in core v4

### ISSUE THREAD

**Issue by SimonCropp** [Particular/NServiceBus.Callbacks#66]
Doing a Request in callbacks v2 and a reply in core v4 results in 

```
System.InvalidOperationException: No handlers could be found for message type: ObjectResponseMessage
   at NServiceBus.LoadHandlersConnector.&lt;Invoke&gt;d__1.MoveNext() in NServiceBus.Core\Pipeline\Incoming\LoadHandlersConnector.cs:line 31
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.DeserializeLogicalMessagesConnector.&lt;Invoke&gt;d__1.MoveNext() in NServiceBus.Core\Pipeline\Incoming\DeserializeLogicalMessagesConnector.cs:line 33
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.SubscriptionReceiverBehavior.&lt;Invoke&gt;d__1.MoveNext() in NServiceBus.Core\Routing\MessageDrivenSubscriptions\SubscriptionReceiverBehavior.cs:line 29
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.ReceivePerformanceDiagnosticsBehavior.&lt;Invoke&gt;d__2.MoveNext() in NServiceBus.Core\Performance\Statistics\ReceivePerformanceDiagnosticsBehavior.cs:line 40
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.ProcessingStatisticsBehavior.&lt;Invoke&gt;d__0.MoveNext() in NServiceBus.Core\Performance\Statistics\ProcessingStatisticsBehavior.cs:line 27
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.TransportReceiveToPhysicalMessageProcessingConnector.&lt;Invoke&gt;d__1.MoveNext() in NServiceBus.Core\Pipeline\Incoming\TransportReceiveToPhysicalMessageProcessingConnector.cs:line 37
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.MainPipelineExecutor.&lt;Invoke&gt;d__1.MoveNext() in NServiceBus.Core\Pipeline\MainPipelineExecutor.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.ReceiveStrategy.&lt;TryProcessMessage&gt;d__7.MoveNext() in NServiceBus.Core\Transports\Msmq\ReceiveStrategy.cs:line 109
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.TransactionScopeStrategy.&lt;ProcessMessage&gt;d__2.MoveNext() in NServiceBus.Core\Transports\Msmq\TransactionScopeStrategy.cs:line 86</Value>
  </HeaderInfo>
```
Adding the handler for ObjectResponseMessage results in the callback never being execited

Repro: https://github.com/SimonCropp/scratch/tree/master/callbacks-v6-v4

Click on SendObjectMessage in the web project. 

Google groups issue: https://groups.google.com/forum/#!topic/particularsoftware/551q9DNTBCQ

Note support for 4.7.x ended in 2016-09-29 https://docs.particular.net/nservicebus/upgrades/supported-versions#nservicebus

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers The following seems to happen:

1. v6 sends a message to 4.7.x
1. 4.7.x replies with a message and seems to set `MessageIntentEnum.Reply` but what goes out is `MessageIntentEnum.Send`
![image](https://user-images.githubusercontent.com/174258/27784605-f26011e6-5fda-11e7-8dcb-9e26d2e8b8b3.png)
![image](https://user-images.githubusercontent.com/174258/27784644-1785303c-5fdb-11e7-87c9-8e796bd74bc0.png)
1. Because the message arrives with intent Send the following code trips over
https://github.com/Particular/NServiceBus.Callbacks/blob/develop/src/NServiceBus.Callbacks/IncomingContextExtensions.cs#L34
since with check for Min supported version to be 4.3 https://github.com/Particular/NServiceBus.Callbacks/blob/develop/src/NServiceBus.Callbacks/IncomingContextExtensions.cs#L49

Since 4.7.x is out of support should we fix the callback and set the min supported version to something higher than 4.7 and release that as a patch? 

thoughts @Particular/nservicebus-maintainers ?

**Comment by danielmarbach** [comment]
As an alternative (maybe even better approach) we could remove the whole version check

**Comment by bording** [comment]
@danielmarbach While NSB 4 is out of support, looking at the callbacks code, it seems like checking for the message intent is doing extra work that isn't needed. If the message came in, it's either going to have data in `lookup` or not, so the end result in the same: data or null.

Unless there's some other reason why checking the message intent is important, I'd say that removing that code is worth doing anyway, and it has the side effect of making it work with 4.7.

**Comment by danielmarbach** [comment]
I agree. I can send in a PR
-- 

Have you tried NServiceBus V6 <https://particular.net/nservicebus-v6> yet?

Learn how to scale applications with microservices and NServiceBus 6
<https://www.pluralsight.com/courses/microservices-nservicebus6-scaling-applications>
.
------------------------------

*Join Udi Dahan's next Advanced Distributed Design Course:*

September 2017 Los Angeles, CA, USA <http://particular.net/adsd>

November 2017 Sydney, Australia <http://particular.net/adsd>
------------------------------

**Comment by danielmarbach** [comment]
Closing in favour of https://github.com/Particular/NServiceBus.Callbacks/issues/67 and https://github.com/Particular/NServiceBus.Callbacks/issues/68

**Comment by danielmarbach** [comment]
Short update:

I investigated whether we can entirely remove the intent check and came to the conclusion that it doesn't work. The problem are send locals. When you register a callback for a send local and the message arrives in the input queue that message already has the correlation id set. When checking for intent send or reply this message would then be treated as the response for the callback. 

Talked this through with @timbussmann and we concluded that it is better to change the logic to the following:

If the sender version is less or equal to 4.7 do not check the intent enum. For SendLocals we assume that you are going to use a version which is supported by the callbacks package which is NSB v6 and higher.

With that we unstuck the user. 

Other things we considered:

- Do not do anything since 4.7 is no longer supported. (Would this be considered a wire compat break?)
- Do not try to lookup the response lookup callback state for SendLocals and check intent Send or Reply for all other messages (would require sophisticated logic to detect send local). Not an option we think
- Add an extension method which allows to bypass the check (this would require sophisticated header transportation between sender and receiver). Not an option we think

We decided to send in a PR to discuss further details.
